File: Decorator/cpp-source/notifiers/DecoratorNotifier.cpp
//
// Created by Pavel Akhtyamov on 2019-04-02.
//

#include <DecoratorNotifier.h>

DecoratorNotifier::DecoratorNotifier(std::shared_ptr<Notifier> notifier): notifier_(std::move(notifier)) {}

void DecoratorNotifier::Notify(const std::string &message) {
  notifier_->Notify(message);
}

File: Decorator/python-source/notifiers/DecoratorNotifier.py
from __future__ import annotations

from .Notifier import Notifier


class DecoratorNotifier(Notifier):
    """Базовая обёртка: держит вложенный отправитель и передаёт ему вызов.

    Сама по себе она ничего не добавляет, и в этом смысл: наследники дописывают
    только своё поведение, а передачу дальше получают готовой.
    """

    def __init__(self, notifier: Notifier) -> None:
        self._notifier = notifier

    def notify(self, message: str) -> None:
        self._notifier.notify(message)
